cmd/kube-proxy/app/server.go

```diff
@@ -913,31 +914,84 @@ func (s *ProxyServer) birthCry() {
 // used for NodePort handling.)
 //
 // The order of precedence is:
-//  1. if bindAddress is not 0.0.0.0 or ::, then it is used as the primary IP.
-//  2. if the Node object can be fetched, then its primary IP is used as the primary IP
-//     (and its secondary IP, if any, is just ignored).
-//  3. otherwise the primary node IP is 127.0.0.1.
+//  1. if bindAddress is valid, it's used as the primary IP
+//  2. Use status.addresses from the Node object (type=InternalIP)
+//  3. Resort to loopback addresses
 //
 // In all cases, the secondary IP is the zero IP of the other IP family.
 func detectNodeIPs(client clientset.Interface, hostname, bindAddress string) (v1.IPFamily, map[v1.IPFamily]net.IP) {
```

Comment:
There are a bunch of changes here and I'm still thinking about them...

> Would it be acceptable to _not_ accept localhost (127.0.0.0/8 or ::1/128) even if explicitly set as bind-address?

I don't think so; it doesn't seem totally implausible that someone might want to run a local-only kubernetes install...

It seems like the problem here is just that most of the people saying `--bind-address 127.0.0.1` don't actually want the behavior that they are requesting.

Reply:
> I don't think so; it doesn't seem totally implausible that someone might want to run a local-only kubernetes install...

LOL, no. Now _I'm_ getting confused about what `--bind-address` means.

OK, so the problem, backward-compatibility-wise, is that `--bind-address` _defines_ what `LoadBalancerSourceRanges` value has the magic "allowFromNode" semantics. So if someone is using `--bind-address 127.0.0.1`, then they can specify `LoadBalancerSourceRanges: ["10.0.0.0/8", "127.0.0.1/32"]` and that means "allow from 10.0.0.0/8 and the node". If we started ignoring `--bind-address 127.0.0.1`, then that would change the interpretation of their existing `LoadBalancerSourceRanges`. :exploding_head: 

Is anyone doing that? :man_shrugging:
Are they bad people if they are? Yes!
Are we bad people for designing the API this way? ... No comment.

Reply:
I will verify, but I think you can't send packets 127.0.0.1 -> loadBalancerIP because that would be a "martian". So, that's why I have claimed that localhost addresses in --bind-address (which isn't a bind address) are useless.

Reply:
Verified: including "127.0.0.0/8" in `loadBalancerSourceRanges` _and_ set `bindAddress: "127.0.0.1"` in the kube-proxy config does _not_ allow access from the node, even for proxy-mode=iptables.

<details><summary>Service manifest</summary>

```yaml
apiVersion: v1
kind: Service
metadata:
  name: alpine
spec:
  loadBalancerSourceRanges:
#  - "192.168.1.0/24"
  - "127.0.0.0/8"
  - "192.168.2.0/24"
  - "fd00::c0a8:100/120"
  - "fd00::c0a8:200/120"
  ipFamilyPolicy: RequireDualStack
  selector:
    app: alpine
  type: LoadBalancer
  allocateLoadBalancerNodePorts: false
  ports:
  - port: 5001
    name: nc
```

</details>

```
# In the kube-proxy log
I0803 13:28:06.834076     359 server_others.go:110] "Detected node IP" address="127.0.0.1"
# from a vm on 192.168.2.0/24
nc 10.0.0.0:5001
alpine-6f86fcb8c4-s8t52
# (works)
# On the node:
nc 10.0.0.0:5001
# (hangs...)
# The problem:
ip ro get 10.0.0.0
10.0.0.0 via 192.168.1.201 dev eth1 src 192.168.1.2 uid 0 
# Linux will not pick the loopback as src.
```

Reply:
The "allowFromNode" (hence the entire --bind-address) thing is _only_ useful if the loadBalancerIP is allocated to an interface on the node. So, the above non-working scenario can be "fixed" with;
```
ip add add 10.0.0.0/32 dev eth0
nc 10.0.0.0:5001
alpine-6f86fcb8c4-7fjr4 # (now works)
```

Reply:
> The "allowFromNode" (hence the entire --bind-address) thing is only useful if the loadBalancerIP is allocated to an interface on the node

Well, it's not that it's only _useful_ in that case, it's that it's only _needed_ in that case. If the load balancer IP _isn't_ assigned to a local interface, then node-to-loadbalancer-IP traffic will actually get sent to the load balancer, and then the load balancer can validate the source IP against its own copy of the service's LoadBalancerSourceRanges. (Or, if the LB doesn't implement LoadBalancerSourceRanges itself, then it would either (a) forward the packet to a different node, preserving the source IP, in which case that second node would see that the node's IP was in LoadBalancerSourceRanges and accept it, or (b) hairpin the packet back to the same node again, in which case it would be forced to masquerade it because of hairpinning, in which case the kube-proxy allowFromNode rule would get hit, and the packet would be accepted.)

(So in the (a) case, this only works if the _real_ node IP is in LoadBalancerSourceRanges, and wouldn't work if you used a fake `--bind-address` and corresponding fake LoadBalancerSourceRange.)

Reply:
I still not fully process all these scenarios, is there something important we should worry about or just an edge case on some weird configuration?